tests/lean/run/structInstFast.lean

```diff
@@ -0,0 +1,68 @@
+import Lean
+
+open Lean Parser Command
+
+def mkIdent' (s : String) (i : Nat) := mkIdent <| Name.mkSimple <| s++s!"{i}"
+
+def mkFieldsStx (type : TSyntax `ident) (s : String) (width : Nat) : MacroM (TSyntax ``structFields) :=
+  go type s width []
+where go (type : TSyntax `ident) (s : String) (width : Nat)
+    (binds : List <| TSyntax ``structExplicitBinder) : MacroM (TSyntax ``structFields) := do
+  match width with
+  | 0 => `(structFields| $binds.reverse.toArray*)
+  | m+1 =>
+    let new ← `(structExplicitBinder| ($(mkIdent' s m) : $type))
+    go type s m (new::binds)
+
+def baseIdent := mkIdent (Name.mkSimple "base")
+
+def baseTypeIdent := mkIdent (Name.mkSimple "Base")
+
+def init (type val : TSyntax `ident) (width : Nat) : MacroM (TSyntax `command) := do
+  let fieldsStx ← mkFieldsStx type "base_" width
+  let vals := Array.mkArray width val
+  `(structure $baseTypeIdent where
+    $fieldsStx:structFields
+    def $baseIdent : $baseTypeIdent := ⟨$vals,*⟩)
+
+def mkStructAndInstStx (type val : TSyntax `ident) (width depth: Nat) : MacroM <| TSyntax `command := do
+  let init ← init type val width
+  go val width depth #[init]
+where go (val : TSyntax `ident) (width depth : Nat) (cmds : Array <| TSyntax `command) :
+    MacroM <| TSyntax `command := do
+  match depth with
+  | 0 =>
+    let cmd : TSyntax `command := ⟨mkNullNode cmds⟩
+    `($cmd:command)
+  | m+1 =>
+    let len := cmds.data.length
+    let newTerm (s : String) := if len = 1 then baseTypeIdent else mkIdent' s (m+1)
+    let newTerm' (s : String) := if len = 1 then baseIdent else mkIdent' s (m+1)
+    let fieldsStx ← mkFieldsStx type (s!"x{m}_") width
+    let nextStruct ←
+      `(structure $(mkIdent' "A" m) extends $(newTerm "A") where
+        $fieldsStx:structFields)
+    let structVals ← (List.range width).mapM fun j =>
+      `(Term.structInstField| $(mkIdent' s!"x{m}_" j):ident := $val)
+    let nextStructInst ←
+      `(def $(mkIdent' "a" m) : $(mkIdent' "A" m) := { $(newTerm' "a"):ident with $structVals.toArray,* })
+    let nextStructInst' ←
+      `(def $(mkIdent' "a'" m) : $(mkIdent' "A" m) := { $(newTerm' "a'"):ident with $structVals.toArray,* })
+    let newCmd : TSyntax `command ←
+      `($nextStruct:command
+        $nextStructInst
+        $nextStructInst')
+    go val width m (cmds.push newCmd)
+
+/-- We build a structure with `depth` nested structures and `width` fields at each level.
+    The structure is initialized with `val` at every field using the `with` construction pattern. -/
+macro "deep_wide_struct_inst_with" type:ident val:ident depth:num width:num : command =>
+  mkStructAndInstStx type val width.getNat depth.getNat
+
+def foo : String := "foo"
+
+deep_wide_struct_inst_with String foo 50 20
+
+/- Structure instances using the `with` pattern should remain fast -/
```

Comment:
Please clarify exactly what this is testing, in particular what "remain" means (are you testing that this PR doesn't make something worse, or guarding against future regressions).

Reply:
If it is guarding against future regressions, did this benchmark change in this PR?

Reply:
It improves and guards against future regression. Though it only applies to the `with` pattern. It might be good to write another test separately for other patterns. 

Documentation improved.

Reply:
Thanks, new documentation is good.